### PR TITLE
Implement `truncate!` function

### PR DESCRIPTION
**Pull request by mofeing**
This PR addresses part of #15

# TODO

- [x] Discuss semantics of canonical-form functions
- [x] Document method
- [x] Tests

### CODE REVIEW

File: src/Ansatz/Chain.jl [hunk 1]
```diff
@@ -187,6 +187,33 @@ function canonize_site!(::Open, tn::Chain, site::Site; direction::Symbol, mode =
     return tn
 end
 
+function truncate!(qtn::Chain, bond; threshold::Union{Nothing,Real} = nothing, maxdim::Union{Nothing,Int} = nothing)
```
Comment (jofrevalles) [inline src/Ansatz/Chain.jl:190]:
```suggestion
function truncate!(qtn::Chain, bond::Union{Tuple{Site, Site}, Vector{Site}}; threshold::Union{Nothing,Real} = nothing, maxdim::Union{Nothing,Int} = nothing)
```
Maybe let's specify the type in `bond`

File: src/Ansatz/Chain.jl [hunk 2]
```diff
@@ -187,6 +187,33 @@ function canonize_site!(::Open, tn::Chain, site::Site; direction::Symbol, mode =
     return tn
 end
```
Comment (jofrevalles) [inline src/Ansatz/Chain.jl:189]:
```suggestion

truncate(tn::Chain, bond; kwargs...) = truncate!(deepcopy(tn), bond; kwargs...)
truncate!(tn::Chain, bond; kwargs...) = truncate!(boundary(tn), tn, bond; kwargs...)

```

**Review by jofrevalles — CHANGES_REQUESTED** [review]

File: src/Ansatz/Chain.jl (hunk 1 above)
Comment (mofeing) [reply src/Ansatz/Chain.jl:190]: It's better not to specify the type. Remember, the type should be just used for multiple dispatch.

File: src/Ansatz/Chain.jl (hunk 2 above)
Comment (mofeing) [reply src/Ansatz/Chain.jl:189]: Why do we need to dispatch on `boundary`?

Comment (jofrevalles) [reply src/Ansatz/Chain.jl:189]: Oops, that was a mistake. I remove that, sorry!

File: test/Ansatz/Chain_test.jl [hunk 3]
```diff
@@ -51,6 +51,24 @@
         @test rightsite(qtn, Site(1)) == Site(2)
     end
 
+    @testset "truncate" begin
+        qtn = Chain(State(), Open(), [rand(2, 2), rand(2, 2, 2), rand(2, 2)])
+        canonize_site!(qtn, Site(2); direction = :right, method = :svd)
+
+        @test_throws ArgumentError truncate!(qtn, [Site(1), Site(2)]; maxdim = 1)
```
Comment (mofeing) [inline test/Ansatz/Chain_test.jl:58]: Why should this fail?

Comment (jofrevalles) [reply test/Ansatz/Chain_test.jl:58]: There are no singular values there to truncate

Comment (mofeing) [reply test/Ansatz/Chain_test.jl:58]: Ah ok. Maybe then `ArgumentError` is not the exception we should throw.

Comment (jofrevalles) [reply test/Ansatz/Chain_test.jl:58]: Do you have any suggestion? I can't seem to think of a better one

Comment (mofeing) [reply test/Ansatz/Chain_test.jl:58]: I think we should throw a custom exception of our own. `NonCanonicalFormException`? `MissingSchmidtCoefficientsException`?

Comment (jofrevalles) [reply test/Ansatz/Chain_test.jl]: Okey, done

**Pull request merged**